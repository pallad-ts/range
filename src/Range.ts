export type Range<T> = Range.Start<T> | Range.End<T> | Range.Full<T>;

export namespace Range {
    export type Full<T> = Start<T> & End<T>;
    export namespace Full {
        export function is<T = any>(value: any): value is Full<T> {
            return value && 'start' in value && 'end' in value;
        }
    }

    export interface Start<T> {
        start: T;
    }

    export namespace Start {
        export function is<T = any>(value: any): value is Start<T> {
            return value && 'start' in value;
        }
    }

    export interface End<T> {
        end: T;
    }

    export namespace End {
        export function is<T = any>(value: any): value is End<T> {
            return value && 'end' in value;
        }
    }

    export const isStart = Start.is;
    export const isEnd = End.is;
    export const isFull = Full.is;

    export function is<T>(value: any): value is Range<T> {
        return Full.is(value) || Start.is(value) || End.is(value);
    }
}