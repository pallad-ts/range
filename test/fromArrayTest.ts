import {Range} from '@src/Range';

describe('fromArray', () => {
    it('fails for empty array', () => {
        expect(() => {
            Range.fromArray([]);
        }).toThrowErrorMatchingSnapshot();
    });

    describe('success', () => {
        it('for single element array', () => {
            expect(Range.fromArray([1]))
                .toEqual(Range.create(1));
        });

        it('for array with 2 elements', () => {
            expect(Range.fromArray([1, 10]))
                .toEqual(Range.create(1, 10));
        });

        it('for array with more than 2 elements', () => {
            expect(Range.fromArray([1, 10, 100]))
                .toEqual(Range.create(1, 10));
        });
    })
})
