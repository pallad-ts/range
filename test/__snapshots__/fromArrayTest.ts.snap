// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromArray fails for empty array 1`] = `"Cannot create range from empty array"`;
